CovertInfiltration/Src/CovertInfiltration/Classes/X2StrategyElement_InfiltrationRewards.uc

```diff
@@ -231,3 +257,215 @@ static protected function bool RewardNotAvaliable(optional XComGameState NewGame
+	local XComGameState_Item ItemState;
+	local X2ItemTemplate ItemTemplate;
+	local X2LootTableManager LootManager;
+	local LootResults LootToGive;
+	local name LootName;
+	local int LootIndex, idx;
+	local string LootString;
+
+	XComHQ = class'UIUtilities_Strategy'.static.GetXComHQ();
+	XComHQ = XComGameState_HeadquartersXCom(NewGameState.ModifyStateObject(class'XComGameState_HeadquartersXCom', XComHQ.ObjectID));
+
+	LootManager = class'X2LootTableManager'.static.GetLootTableManager();
+	LootIndex = LootManager.FindGlobalLootCarrier(RewardState.GetMyTemplate().rewardObjectTemplateName);
+	if (LootIndex >= 0)
+	{
+		LootManager.RollForGlobalLootCarrier(LootIndex, LootToGive);
+	}
+	
+	ItemTemplateManager = class'X2ItemTemplateManager'.static.GetItemTemplateManager();
+
+	// First give each piece of loot to XComHQ so it can be collected and added to LootRecovered, which will stack it automatically
+	foreach LootToGive.LootToBeCreated(LootName)
+	{
+		ItemTemplate = ItemTemplateManager.FindItemTemplate(LootName);
+		ItemState = ItemTemplate.CreateInstanceFromTemplate(NewGameState);
+		XComHQ.PutItemInInventory(NewGameState, ItemState, true);
+	}
+
+	// Then create the loot string for the room
+	for (idx = 0; idx < XComHQ.LootRecovered.Length; idx++)
+	{
+		ItemState = XComGameState_Item(NewGameState.GetGameStateForObjectID(XComHQ.LootRecovered[idx].ObjectID));
+
+		if (ItemState != none)
+		{
+			LootString $= ItemState.GetMyTemplate().GetItemFriendlyName() $ " x" $ ItemState.Quantity;
+
+			if (idx < XComHQ.LootRecovered.Length - 1)
+			{
+				LootString $= ", ";
+			}
+		}
+	}
+	RewardState.RewardString = LootString;
+	
+	// Actually add the loot which was generated to the inventory
+	class'XComGameStateContext_StrategyGameRule'.static.AddLootToInventory(NewGameState);
+}
+
+static function string GetLootTableRewardString(XComGameState_Reward RewardState)
+{
+	`CI_Log("String: '" $ RewardState.RewardString $ "'");
+	return RewardState.RewardString;
+}
+
+static function X2DataTemplate CreateTechInspireRewardTemplate()
+{
+	local X2RewardTemplate Template;
+
+	`CREATE_X2Reward_TEMPLATE(Template, 'Reward_TechInspire');
+
+	Template.SetRewardFn = SetTechInspireReward;
+	Template.GiveRewardFn = GiveTechInspireReward;
+	Template.GenerateRewardFn = GenerateTechInspireReward;
+	Template.GetRewardStringFn = GetTechInspireRewardString;
+	Template.GetRewardPreviewStringFn = GetTechInspireRewardString;
+	Template.GetRewardDetailsStringFn = GetTechInspireRewardDetails;
+	Template.GetRewardImageFn = GetTechInspireRewardImage;
+	Template.GetRewardIconFn = class'X2StrategyElement_DefaultRewards'.static.GetGenericRewardIcon;
+	Template.CleanUpRewardFn = class'X2StrategyElement_DefaultRewards'.static.CleanUpRewardWithoutRemoval;
+
+	return Template;
+}
+
+static function SetTechInspireReward(XComGameState_Reward RewardState, optional StateObjectReference RewardObjectRef, optional int Amount)
+{
+	RewardState.RewardObjectReference = RewardObjectRef;
+}
```

Comment:
Here I would check if `RewardObjectRef` refers to XCGS_Tech (and redscreen if not)